mono/metadata/icall.c

```diff
@@ -1530,19 +1524,13 @@ ves_icall_System_ValueType_Equals (MonoObject *this_obj, MonoObject *that, MonoA
 		}
 		default:
 			if (!values)
-				values = g_newa (MonoObject*, mono_class_num_fields (klass) * 2);
-			o = mono_field_get_value_object_checked (mono_object_domain (this_obj), field, this_obj, error);
-			if (!is_ok (error)) {
-				mono_error_set_pending_exception (error);
-				return FALSE;
-			}
-			values [count++] = o;
-			o = mono_field_get_value_object_checked (mono_object_domain (this_obj), field, that, error);
-			if (!is_ok (error)) {
-				mono_error_set_pending_exception (error);
-				return FALSE;
-			}
-			values [count++] = o;
+				values = g_newa (MonoObjectHandle, mono_class_num_fields (klass) * 2);
+			o = mono_field_get_value_object_checked (mono_handle_domain (this_obj), field, MONO_HANDLE_RAW (this_obj), error);
```

Comment:
Same as `ves_icall_System_ValueType_InternalGetHashCode` - we don't want to allocate coop handles in proportion to a user-provided input.

Again, it's unclear to me that we really need to do the assignment in the first loop.